#371 Integrate Circuit Breaker with Http Client

anilgursel opened the pull request:
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [x] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

anilgursel commented on squbs-httpclient/src/main/scala/org/squbs/httpclient/ClientFlow.scala:
```diff
@@ -67,12 +70,13 @@ object ClientFlow {
   js.Flow[Pair[jm.HttpRequest, T],Pair[Try[jm.HttpResponse], T], jd.HostConnectionPool] = {
     val (cCtx, sSettings) = fromJava(connectionContext, settings)
-    toJava[T](apply[T](name, cCtx, sSettings, env)(system, mat))
+    toJava[T](apply[T](name, cCtx, sSettings, None, env)(system, mat))
```
This is temporary..  I am still working on the Java API.  I need to map 

`japi.CircuitBreakerSettings[jm.HttpRequest, jm.HttpResponse, T]` to `CircuitBreakerSettings[HttpRequest, HttpResponse, T]`.  

japi.CircuitBreakerSettings already has a `toScala`; however, the generic parameters need to be mapped as well.

Working on it..

anilgursel replied:
Also, there will be a builder for Java.  It will look like as follows:

```java
ClientFlow
    .withFallback(...)
    .withFailureDecider(...)
    .create(name, system, mat)
```

anilgursel commented on squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowCircuitBreakerSpec.scala:
```diff
@@ -0,0 +1,293 @@
+    combinedResponses map { _ should contain theSameElementsAs expected }
+  }
+
+  it should "share the circuit breaker state across multiple flows" in {
```
On top of this test, we should also have a similar test where we do not pass `CircuitBreakerSettings` but, it automatically should use the same `CircuitBreakerState` if the passed settings are same. 

The test will be added after caching is implemented.

akara commented on squbs-httpclient/src/test/java/org/squbs/httpclient/ClientFlowTest.java:
```diff
@@ -104,6 +111,96 @@ public void testClientCallHelloService() throws Exception {
+
+    private void assertJmxValue(ObjectName oName, String key, Object expectedValue) throws Exception {
+        Object actualValue = ManagementFactory.getPlatformMBeanServer().getAttribute(oName, key);
+        Assert.assertEquals(expectedValue, actualValue);
```
@anilgursel I think this Codacy warning is easily fixed 😄 I'm looking through the code in the meantime.

akara commented on squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala:
```diff
@@ -0,0 +1,87 @@
+  private[japi] def failureDeciderAsScala[Out](failureDecider: Function[Try[Out], Boolean]): Try[Out] => Boolean =
+    (out: Try[Out]) => failureDecider.asScala.apply(out)
+
+  private[japi] def failureDeciderOptionAsScala[Out](failureDecider: Optional[Function[Try[Out], Boolean]]) =
```
Ditto - type shadowing.

akara commented on squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala:
```diff
@@ -0,0 +1,87 @@
+  private[japi] def fallbackOptionAsScala[In, Out](fallback: Optional[Function[In, Try[Out]]]) =
+    fallback.asScala.map(fallbackAsScala _)
+
+  private[japi] def failureDeciderAsScala[Out](failureDecider: Function[Try[Out], Boolean]): Try[Out] => Boolean =
```
Ditto - type shadowing.

akara commented on squbs-ext/src/test/java/org/squbs/streams/circuitbreaker/CircuitBreakerBidiFlowTest.java:
```diff
@@ -78,16 +70,17 @@ public void testIncrementFailureCountOnCallTimeout() {
-                        system.scheduler(),
                         2,
                         timeout,
                         FiniteDuration.apply(10, TimeUnit.MILLISECONDS),
```
Evidence, rather prefer passing `10L` for Java API. Adding `FiniteDuration` makes the API a bit higher friction for Java users, I think.

anilgursel replied:
No disagreements.  But, please see https://github.com/paypal/squbs/pull/443#discussion_r109429887.  I should probably use the `FiniteDuration.create` API though.

akara commented on squbs-ext/src/test/java/org/squbs/streams/circuitbreaker/CircuitBreakerStateTest.java:
```diff
@@ -0,0 +1,100 @@
+
+public class CircuitBreakerStateTest {
+
+    final ActorSystem system = ActorSystem.create("CircuitBreakerStateTest");
```
Given JUnit creates a new instance of the Test class for every test method and system is final but not static, how did it not break?

anilgursel replied:
Changing it to `private static final`.  As long as Akka can somehow handle it, there is nothing in the test that would conflict with each other (whether single actor system or multiple actor systems, the tests would work).

akara commented on squbs-ext/src/test/java/org/squbs/streams/circuitbreaker/CircuitBreakerBidiFlowTest.java:
```diff
@@ -53,15 +53,7 @@
-                        "  call-timeout = 50 ms\n" +
-                        "  reset-timeout = 20 ms\n" +
-                        "}"));
+    final ActorSystem system = ActorSystem.create("CircuitBreakerBidiFlowTest");
```
Should this not be static, given JUnit's test behavior?

akara commented on squbs-httpclient/src/test/java/org/squbs/httpclient/ClientFlowCircuitBreakerTest.java:
```diff
@@ -0,0 +1,206 @@
+                "  disable-circuit-breaker = true\n" +
+                "}");
+
+    private static final ActorSystem system = ActorSystem.create("ClientFlowCircuitBreakerTest", config);
```
This is more what I expect to see on a JUnit test consisting of multiple tests and one ActorSystem 😄

I mean `static final ActorSystem`.

akara commented on squbs-httpclient/src/test/java/org/squbs/httpclient/ClientFlowTest.java:
```diff
@@ -104,6 +111,96 @@ public void testClientCallHelloService() throws Exception {
+        assertJmxValue(oNameCB, "CallTimeout", "12 seconds");
+        assertJmxValue(oNameCB, "ResetTimeout", "13 minutes");
+        assertJmxValue(oNameCB, "MaxResetTimeout", "14 days");
+        assertJmxValue(oNameCB, "ExponentialBackoffFactor", 16.0);
```
Looking at `assertJMXValue` source, I see `assertEquals(expectedValue, actualValue);` Perhaps we should pass a message describing which object name it is asserting on. Otherwise you'll just get assertion errors with values without meaningful messages what actually failed.

anilgursel replied:
Changing it as below:

```java
private void assertJmxValue(ObjectName oName, String key, Object expectedValue) throws Exception {
        Object actualValue = ManagementFactory.getPlatformMBeanServer().getAttribute(oName, key);
        assertEquals(oName + "." + key, expectedValue, actualValue);
    }
```

Here is a sample assertion failure message:

```
[error] Test org.squbs.httpclient.ClientFlowTest.testAllInputsAPI failed: java.lang.AssertionError: org.squbs.configuration.ClientFlowTest:type=squbs.httpclient,name="javaAllInputsClient".MaxConnections expected:<42> but was:<41>, took 0.003 sec
```

akara commented on squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowCircuitBreakerSpec.scala:
```diff
@@ -0,0 +1,293 @@
+  val port = serverBinding.localAddress.getPort
+}
+
+class ClientFlowCircuitBreakerSpec  extends AsyncFlatSpec with Matchers with BeforeAndAfterAll {
```
One space too many in front of `extends`?

akara commented on squbs-httpclient/src/test/scala/org/squbs/httpclient/ClientFlowCircuitBreakerSpec.scala:
```diff
@@ -0,0 +1,293 @@
+import org.squbs.streams.circuitbreaker.{CircuitBreakerOpenException, CircuitBreakerSettings}
+import org.squbs.testkit.Timeouts.awaitMax
+
+import scala.concurrent.Await
```
Using postfix ops, please add `import scala.language.postfixOps`

anilgursel commented on squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/impl/AtomicCircuitBreakerState.scala:
```diff
@@ -42,70 +43,90 @@ object AtomicCircuitBreakerState {
+  def create(name: String,
+             maxFailures: Int,
+             callTimeout: FiniteDuration,
+             resetTimeout: FiniteDuration,
```
We've discussed about this in the original TimeoutFlow or CircuitBreaker PR.  I just wanted to align with standard Akka Stream stage, e.g., `delay`.  Please see https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala#L744.

akara replied:
Yes, we can do that if it is a conscious decision.

anilgursel commented on squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala:
```diff
@@ -0,0 +1,87 @@
+  import scala.compat.java8.FunctionConverters._
+  import scala.compat.java8.OptionConverters._
+
+  private[japi] def fallbackAsScala[In, Out](
```
Thanks for catching it!  These functions were originally in a different class, which was requiring the type parameters.  Forgot to clean it up while moving.

anilgursel commented on squbs-ext/src/main/scala/org/squbs/streams/circuitbreaker/japi/CircuitBreakerSettings.scala:
```diff
@@ -0,0 +1,87 @@
+    fallback: Function[In, Try[Out]]): In => Try[Out] = (in: In) => fallback.asScala.apply(in)
+
+  private[japi] def fallbackOptionAsScala[In, Out](fallback: Optional[Function[In, Try[Out]]]) =
+    fallback.asScala.map(fallbackAsScala _)
```
Thanks!  Same issue here.  It was in a different class and the functions were looking slightly different.  Compiler was giving errors and was suggesting me to use `_`.  But, then, I simplified the code and also moved things around, so no need for it anymore.

anilgursel commented on squbs-ext/src/test/scala/org/squbs/streams/circuitbreaker/CircuitBreakerStateSpec.scala:
```diff
@@ -0,0 +1,97 @@
+      1,
+      FiniteDuration.apply(50, TimeUnit.MILLISECONDS),
+      FiniteDuration.apply(20, TimeUnit.MILLISECONDS),
+      FiniteDuration.apply(2, TimeUnit.MINUTES),
```
😂 I was doing a search for `Finite.apply` according to your previous comments for Java usage, and saw this and thought what the help I am doing 😄 Had already fixed.

anilgursel commented:
Addressed the review comment for the first batch.. The only discussion point is `FiniteDuration.apply`.  I have changed the code to use `FiniteDuration.create`, this is the Java API.  I prefer this to align with other standard Akka Stream stages.

akara commented:
I'm OK with using `FiniteDuration.apply` as a conscious decision based on the given rationale. Don't worry.

anilgursel commented:
This PR actually addresses https://github.com/paypal/squbs/issues/373 as well.

akara commented on docs/circuitbreaker.md:
```diff
@@ -82,83 +92,90 @@ Source.from(Arrays.asList("a", "b", "c"))
-#### Creating CircuitBreakerState from system configuration
+CircuitBreakerSettings circuitBreakerSettings =
+        CircuitBreakerSettings.<String, String, UUID>create(circuitBreakerState)
+                .withFailureDecider(out -> out.isFailure() || out.get().equals("b"));
```
the `out.get().equals("b")` can be a bit hard to comprehend. You may want to give a bit more context or use an example like `out.get().equals(StatusCodes.INTERNAL_SERVER_ERROR)`

akara commented on docs/circuitbreaker.md:
```diff
@@ -175,108 +192,46 @@ Please note, in many scenarios, the same circuit breaker instance is used across
 ##### Java
 
-The following API is used to pass a uniqueId mapper:
+A `Function[Context, Optional[Any]` can be provided via `withUniqueIdMapper` function:
```
Missing `]`. `Function[Context, Optional[Any]]`

anilgursel replied:
actually it should be Java style 😄 

`Function<Context, Optional<Any>>`

akara commented on docs/httpclient.md:
```diff
@@ -233,4 +252,71 @@ Visibility of the system configuration has utmost importance while trouble shoot
+
+##### Passing CircuitBreakerSettings Programmatically
+
+You can pass a `CircuitBreakerSettings` instance with the programmatic API.  It lets you to pass a custom `CircuitBreakerState` and optional  fallback, failure decider and unique id mapper functions.  If a `CircuitBreakerSettings` instance is passed in programmatically, then the circuit breaker settings in `application.conf` will be ignored.
```
An API is always programmatic 😄. Better say "You can pass a `CircuitBreakerSettings` instance programmatically. The API lets you pass a custom `CircuitBreakerState`...

akara replied:
... then the circuit breaker settings in `application.conf` **are** ignored.
